File: Common/Json/JsonWriter.vb
Namespace Common

    Public Class JsonWriter
        Implements IDisposable

        Public Sub New(ByVal Output As IO.TextWriter, Optional ByVal LeaveOpen As Boolean = False)
            Me.Out = Output
            Me.LeaveOpen = LeaveOpen
            Me.Reset()
        End Sub

        Public Sub New(ByVal StringBuilder As Text.StringBuilder)
            Me.New(New IO.StringWriter(StringBuilder))
        End Sub

        Private Sub WriteEscaped(ByVal S As String)
            Dim PrevStart = 0
            Dim I = 0
            For I = 0 To S.Length - 1
                Dim Ch = S.Chars(I)
                Dim ECh = Ch
                If EscapeDic.TryGetValue(Ch, ECh) Then
                    Me.Out.Write(S.Substring(PrevStart, I - PrevStart))
                    Me.Out.Write("\"c)
                    Me.Out.Write(ECh)
                    PrevStart = I + 1
                ElseIf Char.IsControl(Ch) Then
                    Me.Out.Write(S.Substring(PrevStart, I - PrevStart))
                    Me.Out.Write("\u")
                    Me.Out.Write(Convert.ToString(Strings.AscW(Ch), 16).PadLeft(4, "0"c))
                    PrevStart = I + 1
                End If
            Next
            Me.Out.Write(S.Substring(PrevStart, I - PrevStart))
        End Sub

        Private Sub WriteNewLine()
            Me.Out.WriteLine()
            For I = 0 To Me.CurrentIndent - 1
                Me.Out.Write(Me.IndentString)
            Next
        End Sub

        Private Sub WriteSeparator(Optional ByVal NewLineRequired As Boolean = False)
            If Me.HasValueBefore Then
                Me.Out.Write(","c)
                If Not NewLineRequired And Not Me.MultiLine And Me.AddSpaces Then
                    Me.Out.Write(" "c)
                End If
            End If
            If Me.HasKeyBefore And Me.AddSpaces Then
                Me.Out.Write(" "c)
            End If

            If (Me.MultiLine And Not Me.HasKeyBefore) Or NewLineRequired Then
                If Me.State <> WriterState.Begin Then
                    Me.WriteNewLine()
                End If
            End If
        End Sub

        Public Sub WriteValue(ByVal Value As String, ByVal Quoted As Boolean)
            Verify.False(Me.State = WriterState.End, "Cannot write after write is finished.")
            Verify.True((Me.State = WriterState.Dictionary).Implies(Me.HasKeyBefore), $"Cannot write a value in place of a key in a dictionary. Use {NameOf(Me.WriteKey)} instead.")

            Me.WriteSeparator()

            If Quoted Then
                Me.Out.Write(""""c)
                Me.WriteEscaped(Value)
                Me.Out.Write(""""c)
            Else
                Me.Out.Write(Value)
            End If

            If Me.State = WriterState.Begin Then
                Me.State = WriterState.End
            End If
            Me.HasKeyBefore = False
            Me.HasValueBefore = True
        End Sub

        Public Sub WriteKey(ByVal Name As String)
            Verify.False(Me.State = WriterState.End, "Cannot write after write is finished.")
            Verify.True(Me.State = WriterState.Dictionary, "Cannot write a key outside a dictionary.")
            Verify.False(Me.HasKeyBefore, "Cannot write a key immediately after another.")

            Me.WriteSeparator()

            Me.Out.Write(""""c)
            Me.WriteEscaped(Name)
            Me.Out.Write(""":")

            Me.HasValueBefore = False
            Me.HasKeyBefore = True
        End Sub

        Public Function OpenList(Optional ByVal MultiLine As Boolean = False) As Opening
            Verify.False(Me.State = WriterState.End, "Cannot write after write is finished.")
            Verify.True((Me.State = WriterState.Dictionary).Implies(Me.HasKeyBefore), $"Cannot write a value in place of a key in a dictionary. Use {NameOf(Me.WriteKey)} instead.")

            Dim R = New Opening(Me, "]"c, If(Me.State = WriterState.Begin, WriterState.End, Me.State), Me.MultiLine)

            Me.WriteSeparator(Me.OpeningBraceOnNewLine And MultiLine)
            Me.Out.Write("["c)

            Me.MultiLine = MultiLine
            Me.HasValueBefore = False
            Me.HasKeyBefore = False
            Me.State = WriterState.List

            If MultiLine Then
                Me.CurrentIndent += 1
            End If

            Return R
        End Function

        Public Function OpenDictionary(Optional ByVal MultiLine As Boolean = False) As Opening
            Verify.False(Me.State = WriterState.End, "Cannot write after write is finished.")
            Verify.True((Me.State = WriterState.Dictionary).Implies(Me.HasKeyBefore), $"Cannot write a value in place of a key in a dictionary. Use {NameOf(Me.WriteKey)} instead.")

            Dim R = New Opening(Me, "}"c, If(Me.State = WriterState.Begin, WriterState.End, Me.State), Me.MultiLine)

            Me.WriteSeparator(Me.OpeningBraceOnNewLine And MultiLine)
            Me.Out.Write("{"c)

            Me.MultiLine = MultiLine
            Me.HasValueBefore = False
            Me.HasKeyBefore = False
            Me.State = WriterState.Dictionary

            If MultiLine Then
                Me.CurrentIndent += 1
            End If

            Return R
        End Function

        Private Sub CloseOpening(ByVal ClosingChar As Char, ByVal PreviousState As WriterState, ByVal PreviousMultiline As Boolean)
            Verify.False(Me.HasKeyBefore, "Cannot close while a key is pending its value.")
            Verify.False(Me.State = WriterState.End, "Cannot write after write is finished.")

            If Me.MultiLine Then
                Me.CurrentIndent -= 1
                Me.WriteNewLine()
            End If

            Me.Out.Write(ClosingChar)

            Me.MultiLine = PreviousMultiline
            Me.State = PreviousState
            Me.HasValueBefore = True
        End Sub

        Private Sub Reset()
            Me.State = WriterState.Begin
            Me.HasValueBefore = False
            Me.HasKeyBefore = False
            Me.MultiLine = False
            Me.CurrentIndent = 0
        End Sub

        Private Shared ReadOnly EscapeDic As Dictionary(Of Char, Char) =
                (Function() New Dictionary(Of Char, Char) From {
                         {""""c, """"c},
                         {"/"c, "/"c},
                         {"\"c, "\"c},
                         {Strings.ChrW(&H8), "b"c},
                         {Strings.ChrW(&HC), "f"c},
                         {Strings.ChrW(&HA), "n"c},
                         {Strings.ChrW(&HD), "r"c},
                         {Strings.ChrW(&H9), "t"c}
                     }).Invoke()

        Private ReadOnly Out As IO.TextWriter
        Private ReadOnly LeaveOpen As Boolean

        Private HasValueBefore As Boolean
        Private HasKeyBefore As Boolean
        Private MultiLine As Boolean
        Private CurrentIndent As Integer
        Private State As WriterState = WriterState.Begin

#Region "IDisposable Support"
        Protected Overridable Sub Dispose(Disposing As Boolean)
            If Not Me.IsDisposed Then
                If Disposing Then
                    If Not Me.LeaveOpen Then
                        Me.Out.Dispose()
                    End If
                End If
            End If
            Me._IsDisposed = True
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            Me.Dispose(True)
        End Sub

#Region "IsDisposed Read-Only Property"
        Private _IsDisposed As Boolean

        <ComponentModel.EditorBrowsable(ComponentModel.EditorBrowsableState.Advanced)>
        Public ReadOnly Property IsDisposed As Boolean
            Get
                Return Me._IsDisposed
            End Get
        End Property
#End Region
#End Region

#Region "IndentString Property"
        Private _IndentString As String = "  "

        Public Property IndentString As String
            Get
                Return Me._IndentString
            End Get
            Set(ByVal Value As String)
                Me._IndentString = Value
            End Set
        End Property
#End Region

#Region "OpeningBraceOnNewLine Property"
        Private _OpeningBraceOnNewLine As Boolean = False

        Public Property OpeningBraceOnNewLine As Boolean
            Get
                Return Me._OpeningBraceOnNewLine
            End Get
            Set(ByVal Value As Boolean)
                Me._OpeningBraceOnNewLine = Value
            End Set
        End Property
#End Region

#Region "AddSpaces Property"
        Private _AddSpaces As Boolean = True

        Public Property AddSpaces As Boolean
            Get
                Return Me._AddSpaces
            End Get
            Set(ByVal Value As Boolean)
                Me._AddSpaces = Value
            End Set
        End Property
#End Region

        Public Structure Opening
            Implements IDisposable

            Friend Sub New(ByVal Writer As JsonWriter, ByVal ClosingChar As Char, ByVal PreviousState As WriterState, ByVal PreviousMultiline As Boolean)
                Me.Writer = Writer
                Me.ClosingChar = ClosingChar
                Me.PreviousState = PreviousState
                Me.PreviousMultiline = PreviousMultiline
            End Sub

            Private Sub Dispose() Implements IDisposable.Dispose
                Me.Writer.CloseOpening(Me.ClosingChar, Me.PreviousState, Me.PreviousMultiline)
            End Sub

            Public Sub Close()
                Me.Dispose()
            End Sub

            Private ReadOnly ClosingChar As Char
            Private ReadOnly PreviousState As WriterState
            Private ReadOnly PreviousMultiline As Boolean
            Private ReadOnly Writer As JsonWriter

        End Structure

        Friend Enum WriterState

            Begin
            Dictionary
            List
            [End]

        End Enum

    End Class

End Namespace

File: Common/Json/JsonWriter.cs
using System;
using System.Collections.Generic;

namespace Ks.Common
{
    public class JsonWriter : IDisposable
    {
        public JsonWriter(System.IO.TextWriter Output, bool LeaveOpen = false)
        {
            this.Out = Output;
            this.LeaveOpen = LeaveOpen;
            this.Reset();
        }

        public JsonWriter(System.Text.StringBuilder StringBuilder) : this(new System.IO.StringWriter(StringBuilder))
        {
        }

        private void WriteEscaped(string S)
        {
            var PrevStart = 0;
            var I = 0;
            for (; I < S.Length; I++)
            {
                var Ch = S[I];
                if (EscapeDic.TryGetValue(Ch, out var ECh))
                {
                    this.Out.Write(S[PrevStart..I]);
                    this.Out.Write('\\');
                    this.Out.Write(ECh);
                    PrevStart = I + 1;
                }
                else if (char.IsControl(Ch))
                {
                    this.Out.Write(S[PrevStart..I]);
                    this.Out.Write(@"\u");
                    this.Out.Write(Utilities.Math.ConvertToBase(Ch, 16).PadLeft(4, '0'));
                    PrevStart = I + 1;
                }
            }
            this.Out.Write(S[PrevStart..I]);
        }

        private void WriteNewLine()
        {
            this.Out.WriteLine();
            for (var I = 0; I < this.CurrentIndent; I++)
            {
                this.Out.Write(this.IndentString);
            }
        }

        private void WriteSeparator(bool NewLineRequired = false)
        {
            if (this.HasValueBefore)
            {
                this.Out.Write(',');
                if (!NewLineRequired & !this.MultiLine & this.AddSpaces)
                {
                    this.Out.Write(' ');
                }
            }
            if (this.HasKeyBefore & this.AddSpaces)
            {
                this.Out.Write(' ');
            }

            if ((this.MultiLine & !this.HasKeyBefore) | NewLineRequired)
            {
                if (this.State != WriterState.Begin)
                {
                    this.WriteNewLine();
                }
            }
        }

        public void WriteValue(string Value, bool Quoted)
        {
            Verify.False(this.State == WriterState.End, "Cannot write after write is finished.");
            Verify.True((this.State == WriterState.Dictionary).Implies(this.HasKeyBefore), $"Cannot write a value in place of a key in a dictionary. Use {nameof(this.WriteKey)} instead.");

            this.WriteSeparator();

            if (Quoted)
            {
                this.Out.Write('"');
                this.WriteEscaped(Value);
                this.Out.Write('"');
            }
            else
            {
                this.Out.Write(Value);
            }

            if (this.State == WriterState.Begin)
            {
                this.State = WriterState.End;
            }

            this.HasKeyBefore = false;
            this.HasValueBefore = true;
        }

        public void WriteKey(string Name)
        {
            Verify.False(this.State == WriterState.End, "Cannot write after write is finished.");
            Verify.True(this.State == WriterState.Dictionary, "Cannot write a key outside a dictionary.");
            Verify.False(this.HasKeyBefore, "Cannot write a key immediately after another.");

            this.WriteSeparator();

            this.Out.Write('"');
            this.WriteEscaped(Name);
            this.Out.Write("\":");

            this.HasValueBefore = false;
            this.HasKeyBefore = true;
        }

        public Opening OpenList(bool MultiLine = false)
        {
            Verify.False(this.State == WriterState.End, "Cannot write after write is finished.");
            Verify.True((this.State == WriterState.Dictionary).Implies(this.HasKeyBefore), $"Cannot write a value in place of a key in a dictionary. Use {nameof(this.WriteKey)} instead.");

            var R = new Opening(this, ']', (this.State == WriterState.Begin) ? WriterState.End : this.State, this.MultiLine);

            this.WriteSeparator(this.OpeningBraceOnNewLine & MultiLine);
            this.Out.Write('[');

            this.MultiLine = MultiLine;
            this.HasValueBefore = false;
            this.HasKeyBefore = false;
            this.State = WriterState.List;

            if (MultiLine)
            {
                this.CurrentIndent += 1;
            }

            return R;
        }

        public Opening OpenDictionary(bool MultiLine = false)
        {
            Verify.False(this.State == WriterState.End, "Cannot write after write is finished.");
            Verify.True((this.State == WriterState.Dictionary).Implies(this.HasKeyBefore), $"Cannot write a value in place of a key in a dictionary. Use {nameof(this.WriteKey)} instead.");

            var R = new Opening(this, '}', (this.State == WriterState.Begin) ? WriterState.End : this.State, this.MultiLine);

            this.WriteSeparator(this.OpeningBraceOnNewLine & MultiLine);
            this.Out.Write('{');

            this.MultiLine = MultiLine;
            this.HasValueBefore = false;
            this.HasKeyBefore = false;
            this.State = WriterState.Dictionary;

            if (MultiLine)
            {
                this.CurrentIndent += 1;
            }

            return R;
        }

        private void CloseOpening(char ClosingChar, WriterState PreviousState, bool PreviousMultiline)
        {
            Verify.False(this.HasKeyBefore, "Cannot close while a key is pending its value.");
            Verify.False(this.State == WriterState.End, "Cannot write after write is finished.");

            if (this.MultiLine)
            {
                this.CurrentIndent -= 1;
                this.WriteNewLine();
            }

            this.Out.Write(ClosingChar);

            this.MultiLine = PreviousMultiline;
            this.State = PreviousState;
            this.HasValueBefore = true;
        }

        private void Reset()
        {
            this.State = WriterState.Begin;
            this.HasValueBefore = false;
            this.HasKeyBefore = false;
            this.MultiLine = false;
            this.CurrentIndent = 0;
        }

        private static readonly Dictionary<char, char> EscapeDic = new Dictionary<char, char>()
            {
                {'"', '"'},
                {'/', '/'},
                {'\\', '\\'},
                {(char)0x08, 'b'},
                {(char)0x0C, 'f'},
                {(char)0x0A, 'n'},
                {(char)0x0D, 'r'},
                {(char)0x09, 't'}
            };

        private readonly System.IO.TextWriter Out;
        private readonly bool LeaveOpen;

        private bool HasValueBefore;
        private bool HasKeyBefore;
        private bool MultiLine;
        private int CurrentIndent;
        private WriterState State = WriterState.Begin;

        protected virtual void Dispose(bool Disposing)
        {
            if (!this.IsDisposed)
            {
                if (Disposing)
                {
                    if (!this.LeaveOpen)
                    {
                        this.Out.Dispose();
                    }
                }
            }
            this.IsDisposed = true;
        }

        public void Dispose()
        {
            this.Dispose(true);
        }

        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Advanced)]
        public bool IsDisposed { get; private set; }

        public string IndentString { get; set; } = "  ";

        public bool OpeningBraceOnNewLine { get; set; } = false;

        public bool AddSpaces { get; set; } = true;

        public struct Opening : IDisposable
        {
            internal Opening(JsonWriter Writer, char ClosingChar, WriterState PreviousState, bool PreviousMultiline)
            {
                this.Writer = Writer;
                this.ClosingChar = ClosingChar;
                this.PreviousState = PreviousState;
                this.PreviousMultiline = PreviousMultiline;
            }

            void IDisposable.Dispose()
            {
                this.Close();
            }

            public void Close()
            {
                this.Writer.CloseOpening(this.ClosingChar, this.PreviousState, this.PreviousMultiline);
            }

            private readonly char ClosingChar;
            private readonly WriterState PreviousState;
            private readonly bool PreviousMultiline;
            private readonly JsonWriter Writer;
        }

        internal enum WriterState
        {
            Begin,
            Dictionary,
            List,
            End
        }
    }
}
